modules/relaychain-oracle/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Relaychain Oracle.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+};
+use frame_system::pallet_prelude::*;
+use module_support::ForeignChainStateQuery;
+use sp_runtime::traits::Hash;
+
+mod mock;
+mod tests;
+
+pub use module::*;
+
+#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo)]
+pub enum RawOrigin {
+	RelaychainOracle,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct VerifiableCall<Call, Hash> {
+	dispatchable_call: Call,
+	state_hash: Hash,
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	pub type VerifiableCallOf<T> = VerifiableCall<<T as Config>::VerifiableTask, <T as frame_system::Config>::Hash>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The outer origin type.
+		type Origin: From<RawOrigin>;
+
+		/// Dispatchable task that needs to be verified by oracle for dispatch
+		type VerifiableTask: Parameter
+			+ Dispatchable<Origin = <Self as Config>::Origin, PostInfo = PostDispatchInfo>
+			+ From<frame_system::Call<Self>>
+			+ GetDispatchInfo;
+
+		/// Origin that can dispatch calls that have been verified with foreign state
+		type OracleOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = RawOrigin;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CreateActiveQuery { query_hash: T::Hash, state_hash: T::Hash },
+		CallDispatched { task_result: PostDispatchInfo },
+		CallDispatchFailed,
+	}
+
+	// Tasks to be dispatched if foriegn chain state is valid
+	#[pallet::storage]
+	#[pallet::getter(fn query)]
+	pub type ValidateTask<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, VerifiableCallOf<T>, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		IncorrectStateHash,
+		NoMatchingCall,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn dispatch_task(origin: OriginFor<T>, call_hash: T::Hash, state_hash: T::Hash) -> DispatchResult {
+			ValidateTask::<T>::try_mutate_exists(call_hash, |maybe_verifiable_call| -> DispatchResult {
+				T::OracleOrigin::ensure_origin(origin)?;
+
+				let verifiable_call = maybe_verifiable_call.clone().ok_or(Error::<T>::NoMatchingCall)?;
```

Comment:
```suggestion
				let verifiable_call = maybe_verifiable_call.take().ok_or(Error::<T>::NoMatchingCall)?;
```

and no need to make it None later